walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/servlets/FindMeetingQuery.java

```diff
@@ -0,0 +1,158 @@
+
+/** Class that calculates open slot times given Event/Meeting Requests */
+public final class FindMeetingQuery
+{
+    /** This method requires a collection of Event objects and a MeetingRequest Object. 
+        Returns a collection of TimeRange objects that are potential times for meetings*/
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request)
+    {
+        long meetingDuration = request.getDuration();
+        Collection<TimeRange> availableTime = new ArrayList<TimeRange>(); 
+        if (meetingDuration > TimeRange.WHOLE_DAY.duration())
+        {
+            return Collections.emptyList();
+        }
+        //If there are no meeting attendees or no events to look at 
+        else if ((request.getAttendees().isEmpty()&&request.getOptionalAttendees().isEmpty()) || events.isEmpty())
+        {
+            return Arrays.asList(TimeRange.WHOLE_DAY);
+        }
+        else
+        {
+            for(Event event : events)
+            {
+                TimeRange whenEvent = event.getWhen();
+                Set<String> eventAttendees = event.getAttendees();
+                //Check if event attendees are invited to meeting OR there are no mandotory attendees/only optionals
+                if (checkAttendees(eventAttendees,request.getAttendees())||
+                    (request.getAttendees().isEmpty() && !request.getOptionalAttendees().isEmpty()))
+                {
+                    if (availableTime.isEmpty())
+                    {
+                        addFirstOpenTimes(meetingDuration,availableTime,whenEvent);
+                    }
+                    else
+                    {
+                        checkOverlapsInPotentialTimesWithEventTimes(availableTime, whenEvent, meetingDuration);   
+                    }
+                }  
+                else if (request.getOptionalAttendees().isEmpty())
+                {
+                    return Arrays.asList(TimeRange.WHOLE_DAY);
+                } 
+                //Checks if optional attendees can attend open times, but will not affect times made for mandatory attendees
+                if (checkAttendees(eventAttendees, request.getOptionalAttendees()) && !request.getAttendees().isEmpty())
+                {
+                    return checkIfOptionalAttendeesCanAttend(availableTime,whenEvent);
+                }
+            }
+        }
+        return availableTime;
+    }
+
+    public void addFirstOpenTimes(long meetingDuration, Collection<TimeRange> availableTime, TimeRange whenEvent)
+    {
+        TimeRange frontTime1= TimeRange.fromStartEnd(TimeRange.START_OF_DAY,whenEvent.start(),false);
+        TimeRange backTime1 = TimeRange.fromStartEnd(whenEvent.end(),TimeRange.END_OF_DAY,true);
+        addFrontAndBackTimesToList(meetingDuration, frontTime1, backTime1, availableTime);
+    }
+    
+    public void checkOverlapsInPotentialTimesWithEventTimes(Collection<TimeRange> availableTime,TimeRange whenEvent,
```

Comment:
since this function does a lot of stuff, it would be good to add a doc-comment describing what it does.